anitab-org/bridge-in-tech-backend
# Dev  refactor date fields to a human readable format

### Description
Refactor date fields in BIT API endpoints to accept and view human readable date.
The input fields take the format of YYYY-MM-DD HH:MM where user put the date and time relevant to their local time. The system converts any input to UTC datetime so the database keep all datetime related data in UTC time.
Any user retrieving the data regardless whether the data was created/updated by them will be shown the date and time in their own local time. The return date has the format of YYYY-MM-DD HH:MM Z where Z is the logged in user timezone info.

**IMPORTANT** to test this PR, you must drop the existing database tables and all enum types under the Types directory under bitschema schema. 

This is because:
* the data you have is no longer valid for timezone enum related fields
* I have change the list of enum to reflect the pytz library `all_timezones` list (~600 timezones) to cover as wide range of timezone possible  

Fixes #115 

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
- create the followings:
  * create a user, verify email, login (for this user, I creeated user mtreacy01 and used `Europe/London` as timezone
  * create user additional info with `is_organization_rep` value is True 
- create an organization
    * send GET/organizations to view all organizations in the record. You'll see the organization that you've just created. On successful reesponse, see below. (Notice that the date is a human readable date and it has the logged-in user's timezone info attached to it to avoid ambiguity)
<img width="1576" alt="Screen Shot 2020-08-21 at 10 22 22 am" src="https://user-images.githubusercontent.com/29667122/90839837-d0c18180-e39b-11ea-9b9f-68cd1f9ac29c.png">

- create a program with start and end date using format as mentioned in the `model` of the input fields
  * send GET /organizations/{organization_id}/programs request. On successful process see below.  
<img width="1602" alt="Screen Shot 2020-08-21 at 10 23 05 am" src="https://user-images.githubusercontent.com/29667122/90839977-285fed00-e39c-11ea-9491-a3357f4c1723.png">
<img width="1562" alt="Screen Shot 2020-08-21 at 10 23 21 am" src="https://user-images.githubusercontent.com/29667122/90839985-2dbd3780-e39c-11ea-94b3-6703dce01755.png">

To check that other user see the dates in their own timezone interpretation, do the followings:
- create another user (for this I created user ksmith01 and used `Australia/Melbourne` as timezone
- login and send GET /organizations request. On successful process see below. Notice that the creation_date was shown in ksmith timezone:
<img width="1565" alt="Screen Shot 2020-08-21 at 10 24 20 am" src="https://user-images.githubusercontent.com/29667122/90839600-47aa4a80-e39b-11ea-9235-116a3ed839b6.png">

- send GET/organizations/{organization_id}/programs. On successful process see below (again, noticed that all dates are shown in ksmith timezone, not in mtreacy's timezone whose the owner of the organization and program data)
<img width="1571" alt="Screen Shot 2020-08-21 at 10 24 49 am" src="https://user-images.githubusercontent.com/29667122/90839823-c30bfc00-e39b-11ea-8532-438e8cdd5102.png">
<img width="1570" alt="Screen Shot 2020-08-21 at 10 25 02 am" src="https://user-images.githubusercontent.com/29667122/90839827-c7381980-e39b-11ea-9870-823177490801.png">




### Checklist:


**Code/Quality Assurance Only**
